test_runner/regress/test_vm_bits.py

```diff
@@ -117,7 +117,67 @@ def test_vm_bit_clear(neon_simple_env: NeonEnv):
 # Test that the ALL_FROZEN VM bit is cleared correctly at a HEAP_LOCK
 # record.
 #
-def test_vm_bit_clear_on_heap_lock(neon_env_builder: NeonEnvBuilder):
+# This is a repro for the bug fixed in commit 66fa176cc8.
+#
+def test_vm_bit_clear_on_heap_lock_whitebox(neon_env_builder: NeonEnvBuilder):
+    env = neon_env_builder.init_start()
+
+    endpoint = env.endpoints.create_start(
+        "main",
+        config_lines=[
+            # If auto-analyze runs at the same time that we run VACUUM FREEZE, it
+            # can hold a snasphot that prevent the tuples from being frozen.
+            "autovacuum=off",
+            "log_checkpoints=on"
+        ],
+    )
+
+    pg_conn = endpoint.connect()
+    cur = pg_conn.cursor()
+
+    # Install extension containing function needed for test
+    cur.execute("CREATE EXTENSION neon_test_utils")
+    cur.execute("CREATE EXTENSION pageinspect")
+
+    # Create a test table and freeze it to set the all-frozen VM bit on all pages.
+    cur.execute("CREATE TABLE vmtest_lock (id integer PRIMARY KEY)")
+    cur.execute("INSERT INTO vmtest_lock SELECT g FROM generate_series(1, 50000) g")
+    cur.execute("VACUUM (FREEZE, DISABLE_PAGE_SKIPPING true, VERBOSE) vmtest_lock")
+    for notice in pg_conn.notices:
+        log.info(f"{notice}")
+
+    # Lock a row. This clears the all-frozen VM bit for that page.
+    cur.execute("BEGIN")
+    cur.execute("SELECT * FROM vmtest_lock WHERE id = 40000 FOR UPDATE")
+    cur.execute("COMMIT")
+
+    # The VM page in shared buffer cache, and the same page as reconstructed by
+    # the pageserver, should be equal. Except for the LSN: Clearing a bit in the
+    # VM doesn't bump the LSN in PostgreSQL, but the pageserver updates the LSN
+    # when it replays the VM-bit clearing record (since commit 387a36874c)
+    #
+    # This is a bit fragile, we've had lot of flakiness in this test before. For
+    # example, because all the VM bits were not set because concurrent
+    # autoanalyze prevented the VACUUM FREEZE from freezing the tuples. Or
+    # because autoavacuum kicked in and re-froze the page between the
+    # get_raw_page() and get_raw_page_at_lsn() calls. We disable autovacuum now,
+    # which should make this deterministic.
+    cur.execute("select get_raw_page( 'vmtest_lock', 'vm', 0 )")
+    vm_page_in_cache = (cur.fetchall()[0][0])[8:100].hex()
+    cur.execute(
+        "select get_raw_page_at_lsn( 'vmtest_lock', 'vm', 0, pg_current_wal_insert_lsn(), NULL )"
+    )
+    vm_page_at_pageserver = (cur.fetchall()[0][0])[8:100].hex()
```

Comment:
Why the range 8:100? It is where the `ALL_FROZEN` bit resides?

Reply:
No, first 24 bytes of page is page header. First 8 bytes is page LSN. 
The question is whether it is enough or not. I saw some errors when zero page (page containing all zeros) is compared with empty page (page just initialized, but containing no data).

Reply:
> No, first 24 bytes of page is page header. First 8 bytes is page LSN.

I see. Per <https://www.postgresql.org/docs/current/storage-page-layout.html> we are inspecting 19 itemids (and the flags).

> The question is whether it is enough or not. I saw some errors when zero page (page containing all zeros) is compared with empty page (page just initialized, but containing no data).

I wonder if this means that free space part can contain garbage? What if we masked out the free space and then compared the whole pages? (I lost the python code for this but it's easy to unpack the page header with `struct.unpack` and zero out the part between `lower:upper`. Can write it again if interesting.

Reply:
The reason we compare only the first 100 bytes is that the assertion failure becomes unwieldy if the whole page is incuded. All the interesting bytes are in the beginning of the page, the rest should be all zeros. If we had a bug where there was some garbage at the end of the page, we would miss it. But that's not how the original bug would manifest, and doesn't seem like a very likely failure mode anyway.

A more proper way of doing this would be to add a second argument to the assertion, with a custom error message that would print the pages in a nicer format on assertion failure. Can I nerd-snipe someone to write that? ;-)

Reply:
> A more proper way of doing this would be to add a second argument to the assertion, with a custom error message that would print the pages in a nicer format on assertion failure. Can I nerd-snipe someone to write that? ;-)

Looking at Christian's review comment I think the comparison error exception is just fine. I am increasingly confused about the `8:100` ... If that is not `8:` because of assertion messages, shouldn't we just chunk the `8:` to be 16 or 32 bytes and check those individually? Chunking & asserting would cause us to lose if the pages differ in two places.. Maybe doing a `diff -u <(hexdump -Cv --skip 8 page_from_cache) <(hexdump -Cv --skip 8 page_from_ps)` would be easiest and cleanest for the assertion message, and assert on the exit code?

Reply:
Opened https://github.com/neondatabase/neon/issues/8021 to track improving this